### PR TITLE
Enable Login/logout Multitenancy inside of Elsa Workflow Dashboard  

### PR DESCRIPTION
**Pull request by uppercuut**
#3254 

First Things First. 

**### Authentication :** 
To authenticate , we can do that using 3 methods :
1. -open Id Connect 
2. -Server cookies  
3. -jwt bearer 

### **1.OpenIdConnect**
we can use that by simply call new extension service method 

`AddElsaOpenIdConnect` ==> `ElsaOpenIdConnectOptions` 

**mechanism** :  simply calls an extremal IDP server that returns with the authentication.
**brief** : thin layer on top of the normal  `AddOpenIdConnect` top wrap some complexity for people who are only focused on Elsa workflow mainly and register a few  configurations that can help us at the front end.

### **2.Cookies**
`AddElsaCookieAuthentication` ==> `ElsaCookieOptions` 

**mechanism** :  if you like to make your own account/login page and redirect to it then inside of the action you call 
`httpContext.SignInAsync()` with you own claims and data u can using this method.
**brief** : thin layer on top of the normal  `AddCookie` top wrap some complexity for people who are only focused on Elsa workflow mainly and register a few  configurations that can help us at the front end.

### **3.Jwt Bearer**
`AddElsaJwtBearerAuthentication` ==> `ElsaJwtOptions` 

**mechanism** :  an access token managed on the client side and sent in header with each request.(yet to be done at the front end. )
**brief** : thin layer on top of the normal  `ElsaJwtOptions` top wrap some complexity for people who are only focused on Elsa workflow mainly and register a few  configurations that can help us at the front end.

sample on an **AddElsaOpenIdConnect:**

```
            services.AddElsaOpenIdConnect(OpenIdConnectDefaults.AuthenticationScheme, options =>
            {
                options.Authority = "https://localhost:44318/";
                options.ClientId = "ElsaDashboardClientServer";
                options.ResponseType = "code";
                options.Scopes.Add("openid");
                options.Scopes.Add("profile");
                options.Scopes.Add("address");
                options.Scopes.Add("TenantId");
                options.ClientSecret = "Elsa";
                options.UniqueJsonKeys.Add("TenantId", "TenantId");
                options.GetClaimsFromUserInfoEndpoint = true;
            });
```

### **Multitenancy** 

New To methods Called : 
```

AddTenantAccessorFromClaim
AddTenantAccessorFromHeader
```
as the name suggests those are implementation of the interface  `ITenantAccessor` in which I can the implementation that  to that to get the TenantId From 

The Front-end : 

image

I have added a simple component called `elsa-user-context-menu` which contain a button displaying the current user's name 
and a dropdown list has only one option -for time being- which is logout. 

**### Controllers :** 

 Three Controllers were added: 
 **1-log out Controller =>> used for log out users/clear cookies** 
 **2-user details getting the data from claims currently returning only name/tenanted if there are any** 
3-**ElsaAuthenticationContextController => this controller is used to tell the rest of the application and the client side about the configurations regarding the Authentication like : **  

What are the authentication styles and based on that we can know whether we clean the cookies on out server (cookies ). 
or just clean the cookies on the client side (Jwt case) 
or remove the cookies  from the IDP server. 
-we can build a micro engine to handle different cases at the front-end using based on those configuration-

plus the Tenant access how out, client side  can check if it must include the tenantId in the header for example. 

this is pretty much it. 
hope you find it useful. 

please let me know if  there are any notes and i will fix them asap. Much love

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/elsa-workflows/elsa-core?pullRequest=3284) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/elsa-workflows/elsa-core?pullRequest=3284) before we can accept your contribution.<br/><br/>:x: uppercuut [sign now](https://cla.dotnetfoundation.org/elsa-workflows/elsa-core?pullRequest=3284)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/elsa-workflows/elsa-core?pullRequest=3284) it.</sub>

**Comment by Daxton47** [comment]
This is awesome!! Are there plans to get this merged into Elsa 3.0 eventually as well?

**Comment by uppercuut** [comment]
> This is awesome!! Are there plans to get this merged into Elsa 3.0 eventually as well?

Generally, I can put some time/work into elsa3 for this and other work

File: src/server/Elsa.Server.Authentication/TenantAccessors/TenantAccessorFromClaim.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Elsa.Server.Authentication.TenantAccessors
+{
+    public class TenantAccessorFromClaim : ITenantAccessor
+    {
+        private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly string _calimName;
```
Comment (sfmskywalker) [inline src/server/Elsa.Server.Authentication/TenantAccessors/TenantAccessorFromClaim.cs:14]: Small typo (`_calimName` -> `claimName`).

**Review by sfmskywalker — APPROVED** [review]
Looks great, thanks for the contribution!

**Comment by uppercuut** [comment]
Hello, any chance this will be merged soon ?

**Comment by sfmskywalker** [comment]
@uppercuut Quick question about the Dashboard - will it still work out of the box when users do not enable your authentication module?

**Comment by uppercuut** [comment]
yes I'm pretty sure i tested it, but i was a while a go I can recheck later today

**Comment by sfmskywalker** [comment]
Awesome, thank you. Would you be interested in showing off your work on tomorrow's community meeting?

**Comment by uppercuut** [comment]
Hello, when is the meeting? I hope I didn't miss it

**Comment by sfmskywalker** [comment]
It was supposed to happen yesterday but I got home late. I’m gonna try and find another evening to host the weekly meetings, Tuesdays have become a bit busy lately.

**Comment by uppercuut** [comment]
@sfmskywalker 
hello , I have checked when the authentication module is not used. every thing looks good and working as expected. 
I made a few changes/optimizations  of the code used callback functions instead of eval

**Comment by uppercuut** [comment]
> It was supposed to happen yesterday but I got home late. I’m gonna try and find another evening to host the weekly meetings, Tuesdays have become a bit busy lately.

no problem tell me when you are available

**Review by sfmskywalker — APPROVED** [review]

**Pull request merged**